File: app/src/main/java/fr/socket/florian/dhome/view/login/ContainerViewHandler.java
package fr.socket.florian.dhome.view.login;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.util.AttributeSet;
import android.view.KeyEvent;

public class ContainerViewHandler extends ConstraintLayout {

    private boolean isKeyboardShown;
    private OnKeyboardStateChange listener;

    public ContainerViewHandler(Context context) {
        super(context);
    }

    public ContainerViewHandler(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ContainerViewHandler(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public void setKeyboardStateListener(OnKeyboardStateChange listener) {
        this.listener = listener;
    }

    @Override
    public boolean dispatchKeyEventPreIme(@NonNull KeyEvent event) {
        if (event.getKeyCode() == KeyEvent.KEYCODE_BACK && isKeyboardShown) {
            // Keyboard is hiding
            isKeyboardShown = false;
            listener.onKeyboardHide();
        }
        return super.dispatchKeyEventPreIme(event);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        final int proposedHeight = MeasureSpec.getSize(heightMeasureSpec);
        final int actualHeight = getHeight();
        if (actualHeight > proposedHeight && !isKeyboardShown) {
            isKeyboardShown = true;
            listener.onKeyboardShow();
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    // Callbacks
    public interface OnKeyboardStateChange {
        void onKeyboardShow();

        void onKeyboardHide();
    }
}

File: app/src/main/java/fr/socket/florian/dhome/view/login/ContainerViewHandler.kt
package fr.socket.florian.dhome.view.login

import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.View
import androidx.constraintlayout.widget.ConstraintLayout

class ContainerViewHandler : ConstraintLayout {

    private var isKeyboardShown: Boolean = false
    private var listener: OnKeyboardStateChange? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    fun setKeyboardStateListener(listener: OnKeyboardStateChange) {
        this.listener = listener
    }

    override fun dispatchKeyEventPreIme(event: KeyEvent): Boolean {
        if (event.keyCode == KeyEvent.KEYCODE_BACK && isKeyboardShown) {
            // Keyboard is hiding
            isKeyboardShown = false
            listener!!.onKeyboardHide()
        }
        return super.dispatchKeyEventPreIme(event)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val proposedHeight = View.MeasureSpec.getSize(heightMeasureSpec)
        val actualHeight = height
        if (actualHeight > proposedHeight && !isKeyboardShown) {
            isKeyboardShown = true
            listener!!.onKeyboardShow()
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    // Callbacks
    interface OnKeyboardStateChange {
        fun onKeyboardShow()

        fun onKeyboardHide()
    }
}
